csharp/src/Ice/IncomingResponseFrame.cs

```diff
@@ -13,105 +13,155 @@ public enum ResultType : byte { Success, Failure };
+                {
+                    Debug.Assert(false);
+                    throw new InvalidDataException(
+                        $"received ice1 response frame with unknown reply status `{ReplyStatus}'");
+                }
+            }
         }
 
-        /// <summary>Starts reading the result carried by this response frame.</summary>
-        /// <returns>The type of the result, either Success or Failure, and an InputStream iterator over this result.
-        /// </returns>
-        public (ResultType ResultType, InputStream InputStream) ReadResult()
+        /// <summary>Reads an empty return value from the response frame.</summary>
+        public void ReadVoidReturnValue()
         {
-            InputStream.StartEncapsulation();
             switch (ReplyStatus)
             {
                 case ReplyStatus.OK:
-                    return (ResultType.Success, InputStream);
+                {
+                    var istr = new InputStream(_communicator, Payload, 1);
+                    istr.StartEncapsulation();
+                    istr.EndEncapsulation();
+                    break;
+                }
                 case ReplyStatus.UserException:
-                    return (ResultType.Failure, InputStream);
-                default:
-                    throw new InvalidOperationException(); // TODO: better exception and message
+                {
+                    var istr = new InputStream(_communicator, Payload, 1);
+                    istr.StartEncapsulation();
+                    RemoteException ex = istr.ReadException();
+                    istr.EndEncapsulation();
+                    throw ex;
+                }
+                case ReplyStatus.ObjectNotExistException:
+                case ReplyStatus.FacetNotExistException:
+                case ReplyStatus.OperationNotExistException:
+                {
+                    throw ReadDispatchException();
+                }
+                case ReplyStatus.UnknownException:
+                case ReplyStatus.UnknownLocalException:
+                case ReplyStatus.UnknownUserException:
+                {
+                    throw ReadUnhandledException();
+                }
             }
         }
 
-        /// <summary>Reads the return value carried by this response frame using the provided InputStreamReader.
-        /// If the response frame carries a failure, ReadReturnValue reads and throws this exception.</summary>
-        /// <param name="reader">InputStreamReader that reads the return value.</param>
-        /// <returns>The return value.</returns>
-        public T ReadReturnValue<T>(InputStreamReader<T> reader)
+        public IncomingResponseFrame(Communicator communicator, ArraySegment<byte> payload)
         {
-            if (ReadResult().ResultType == ResultType.Failure)
+            _communicator = communicator;
+            byte replyStatus = payload[0];
+            if (replyStatus < 0 || replyStatus > 7)
```

Comment:
I think the other way around would be more readable, i.e. first convert the byte to a ReplyStatus, then make sure it's in range. Also since byte is unsigned in C#, `replyStatus < 0` isn't meaningful, it's strange you didn't get a warning.

Reply:
the `<` `>` operators doesn't work on enums, this will require to cast back the enum to a byte to compare it.

Reply:
Removed `< 0` comparison